### PR TITLE
feat: added timeplus output

### PR DESCRIPTION
**Pull request by ye11ow**
This is the output component for [Timeplus](https://www.timeplus.com/) via [ingest api](https://docs.timeplus.com/ingest-api). It supports both Cloud and self-hosted version of Timeplus Enterprise.

The input component is still WIP, I'll submit a separate PR for it.

### CODE REVIEW

File: internal/impl/timeplus/output.go [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+
+	"github.com/redpanda-data/benthos/v4/public/service"
+	"github.com/redpanda-data/connect/v4/internal/impl/timeplus/http"
+)
+
+var outputConfigSpec *service.ConfigSpec
+
+func init() {
+	// TODO: add Version
+	outputConfigSpec = service.NewConfigSpec().
```
Comment (ye11ow) [inline internal/impl/timeplus/output.go:16]: What would be the `Version` here? Shall I just use the next minor version (`4.38.0`)? 🤔

**Comment by ye11ow** [comment]
The [CI](https://github.com/redpanda-data/connect/actions/runs/11075392917/job/30776244275?pr=2900) failed due to 
>Stale go.{mod,sum} detected. This can be fixed with 'make deps'.

I tried to run `make deps` but nothing changes, any suggestion? Thanks!

**Comment by Jeffail** [comment]
Hey @ye11ow don't worry about the go.mod complaints, there's a couple of linting errors worth looking at though:

```
  Warning: exported: exported const TargetTimeplus should have comment (or a comment on this block) or be unexported (revive)
  Warning: exported: exported type Client should have comment or be unexported (revive)
  Warning: exported: exported function NewClient should have comment or be unexported (revive)
  Warning: exported: exported type Writer should have comment or be unexported (revive)
```

File: internal/impl/timeplus/output.go (hunk 1 above)
Comment (Jeffail) [reply internal/impl/timeplus/output.go:16]: Don't worry about adding it, it's being deprecated soon as we're recording the versions separately going forwards.

Comment (ye11ow) [reply internal/impl/timeplus/output.go:16]: Gotcha, thanks!

File: internal/impl/timeplus/output.go [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+}
+
+type timeplus struct {
+	logger *service.Logger
+	client Writer
+}
+
+// Close implements service.Output
+func (t *timeplus) Close(context.Context) error {
+	// TODO: shall we wait for ongoing writes?
```
Comment (mihaitodor) [inline internal/impl/timeplus/output.go:92]: I'd just get rid of `t.client = nil` here and, also the `t.client == nil` check in `Connect`. You'd get a panic if `service.RegisterBatchOutput` fails, so `t.client` can't be `nil`.

File: internal/impl/timeplus/output.go [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+func (t *timeplus) WriteBatch(ctx context.Context, b service.MessageBatch) error {
+	if len(b) == 0 {
+		return nil
+	}
+
+	cols := []string{}
+	rows := [][]any{}
+
+	// Here we assume all messages have the same structure, same keys
+	// Currently we will just skip the message if it is invalid. Probably we should add a feature flag to not skip them by propagating the error
```
Comment (mihaitodor) [inline internal/impl/timeplus/output.go:115]: Normally, Connect outputs return an error if a message fails. This enables users to leverage a `fallback` output so they can implement a dead letter queue for example. They can use a `drop` output in there if they genuinely want to drop failed messages.

**Review by mihaitodor — CHANGES_REQUESTED** [review]
Thanks @ye11ow for working on this! It looks good overall, but there are some small issues which I think should be addressed before merging.

**Comment by ye11ow** [comment]
> Thanks @ye11ow for working on this! It looks good overall, but there are some small issues which I think should be addressed before merging.

Thank you @mihaitodor for such detailed comments! 😄  I'll address them.

File: internal/impl/timeplus/interface.go [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package timeplus
+
+import "context"
+
+// Writer is the interface. Currently only http writer is implemented. Caller needs to make sure all write consistent the same `cols`
```
[the remark this answers is not in the document]
Comment (ye11ow) [reply internal/impl/timeplus/interface.go]: Nice guess! 😄  Yeah, it should be `all writes contain the same cols`. Updated

File: internal/impl/timeplus/http/client.go [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+	}
+
+	logger = logger.With("target", TargetTimeplusd).With("host", ingestURL.Host).With("ingest_url", ingestURL.RequestURI())
+	logger.Info("timeplus http client created")
+
+	return &Client{
+		logger,
+		ingestURL,
+		header,
+		&http.Client{},
```
[the remark this answers is not in the document]
Comment (ye11ow) [reply internal/impl/timeplus/http/client.go]: Updated, used a customized http client

File: internal/impl/timeplus/http/client.go [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+		Columns: cols,
+		Data:    rows,
+	}
+
+	payloadBytes, err := json.Marshal(payload)
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest(http.MethodPost, c.ingestURL.String(), bytes.NewBuffer(payloadBytes))
```
[the remark this answers is not in the document]
Comment (ye11ow) [reply internal/impl/timeplus/http/client.go]: Makes sense. Updated!

File: internal/impl/timeplus/output.go [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+func init() {
+	// TODO: add Version
+	outputConfigSpec = service.NewConfigSpec().
+		Categories("Services").
+		Summary("Sends message to a Timeplus Enterprise stream via ingest endpoint").
+		Description(`
+This output can send message to Timeplus Enterprise Cloud, Timeplus Enterprise (self-hosted) or directly to timeplusd.
+
+This output accepts structured message only. It also expects all message contains the same keys and matches the schema of stream. If the upstream source returns
+unstructured message such as string, please include a processors to contruct the strcuture message. Here is a short sample:
```
[the remark this answers is not in the document]
Comment (ye11ow) [reply internal/impl/timeplus/output.go]: sharp 👁️ ! updated

File: internal/impl/timeplus/output.go [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+A sample config to send the data to timeplusd
+` + "```yml" + `
+output:
+  timeplus:
+    url: http://localhost:3218
+    stream: mystream
+    username: username
+    password: pw
+` + "```" + `
+`)
```
[the remark this answers is not in the document]
Comment (ye11ow) [reply internal/impl/timeplus/output.go]: updated by moving examples from `Description` to `Example`

File: internal/impl/timeplus/output.go [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+output:
+  timeplus:
+    workspace: my_workspace_id
+    stream: mystream
+    apikey: <Api key genereated on web console>
+
+	processors:
+	  - mapping: |
+        root = {}
+        root.raw = content().string() # Make sure your stream contains a string type raw field
```
[the remark this answers is not in the document]
Comment (ye11ow) [reply internal/impl/timeplus/output.go]: fixed

File: internal/impl/timeplus/output.go [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+	if batchPolicy, err = conf.FieldBatchPolicy("batching"); err != nil {
+		return
+	}
+	if maxInFlight, err = conf.FieldMaxInFlight(); err != nil {
+		return
+	}
+
+	out = &timeplus{
+		logger: logger,
+		client: http.NewClient(logger, maxInFlight, target, baseURL, workspace, stream, apikey, username, password),
```
[the remark this answers is not in the document]
Comment (ye11ow) [reply internal/impl/timeplus/output.go]: I thought each output is responsible to respect this settings. Glad that the framework has done so. Removed.

File: internal/impl/timeplus/output.go (hunk 2 above)
Comment (ye11ow) [reply internal/impl/timeplus/output.go]: updated.

File: internal/impl/timeplus/output.go (hunk 3 above)
Comment (ye11ow) [reply internal/impl/timeplus/output.go]: Makes sense. Updated. Now it will just propagate the error.

**Comment by ye11ow** [comment]
@mihaitodor Fixed all the issues. Could you please check again? Thanks again for all the comments!

**Review by mihaitodor — APPROVED** [review]
LGTM, thank you! Just needs a `go mod tidy` to pass the CI.

**Comment by ye11ow** [comment]
> LGTM, thank you! Just needs a `go mod tidy` to pass the CI.

Thank you @mihaitodor for taking care of that for me! I'll pay more attention on this next time 😄

**Pull request merged**